File: python3/apixcli.py
#!/usr/bin/env python3
"""Unofficial Apix API CLI client for making API requests manually"""

import sys
import argparse

import requests
import apix_public_api

from getpass import getpass

parser = argparse.ArgumentParser(
    description='Unofficial Apix API client implementation')
parser.add_argument('-method', "-m", action='store', required=True,
                    help="API method", choices=["AuthenticateByUser","SendInvoiceZIP","DeliveryMethod","AddressQuery","SendPrintZIP","SendPayslip","ListInvoiceZIPs"])

parser.add_argument('-software_name', '-s', required=False,
                    nargs='?', help='software_name')
parser.add_argument("-software_version", '-v', required=False,
                    nargs='?', help='software_version')
parser.add_argument("-transfer_id", '-i', required=False,
                    nargs='?', help='transfer_id')
parser.add_argument("-transfer_key", '-k', required=False,
                    nargs='?', help='transfer_key')

parser.add_argument("-username", '-u', required=False,
                    nargs='?', help='transfer_id')

parser.add_argument("-file", '-f', required=False, nargs='?', help='Payload')

group = parser.add_mutually_exclusive_group(required=True)

group.add_argument('-environment', '-e', action='store', help="API environment",
                   default="SendInvoiceZIP", choices=["test", "prod"])


def main():
    args = parser.parse_args()

    exitcode = 0

    with apix_public_api.ApixPublicApi(args.transfer_id, args.transfer_key, environment=args.environment) as apix:
        response: requests.Response = None

        data = None
        if args.file is not None:
            with open(args.file, "rb") as fh:
                data = fh.read()

        if args.method == "AuthenticateByUser":
            password = getpass()
            response = apix.AuthenticateByUser(args.username, password, environment=args.environment)
        elif args.method == "SendInvoiceZIP":
            response = apix.SendInvoiceZIP(
                args.software_name, args.software_version, data)
        elif args.method == "DeliveryMethod":
            response = apix.DeliveryMethod(data)
        elif args.method == "AddressQuery":
            response = apix.AddressQuery(data)
        elif args.method == "SendPrintZIP":
            response = apix.SendPrintZIP(
                args.software_name, args.software_version, data)
        elif args.method == "SendPayslip":
            response = apix.SendPayslip(data)
        elif args.method == "ListInvoiceZIPs":
            response = apix.ListInvoiceZIPs()
        else:
            raise Exception(
                "Unknown API method: {method}".format(method=args.method))

        if response.status_code != 200:
            raise Exception("Unknown HTTP error: [{status_code}] {content}".format(status_code=response.status_code, content=response.content))
        else:
            if response.content.find(b"<Status>OK</Status>") >= 0:
                print(response.content)
                sys.stderr.write("Request handled succesfully\n")
                exitcode = 0
            else:
                exitcode = 1
                print(response.content)
                sys.stderr.write(
                    "Request was rejected (attach api response for possible questions for Apix) \n")

    sys.exit(exitcode)


if __name__ == '__main__':
    main()

File: perl/apix-cli.pl
#! /bin/env perl

use strict;
use warnings;
use utf8;

# ugly way to import
use lib "./";
use ApixPublicApi;

use Getopt::Std;
use Getopt::Long;

#for Getopt
our %CMDARGS = ();

GetOptions(
    "help|h"               => \&getopt_handler,
    "software_name|s=s"    => \&getopt_handler,
    "software_version|v=s" => \&getopt_handler,
    "transfer_id|i=s"      => \&getopt_handler,
    "transfer_key|k=s"     => \&getopt_handler,
    "file|f=s"             => \&getopt_handler,
    "endpoint|u=s"         => \&getopt_handler,
    "environment|e=s"      => \&getopt_handler,
);

############### __main__ ###############

check_help() or exit(1);

eval { main(); };
if ($@) {
    die($@);
}

sub main {

    my $data = "";
    open(my $fh, "<", $main::CMDARGS{"file"}) or die(sprintf("Error opening %s: %s", $main::CMDARGS{"file"}, $!));
    while (<$fh>) {
        $data .= $_;
    }
    close($fh);

    my $url = undef;
    if ($main::CMDARGS{"endpoint"}) {
        $url = $main::CMDARGS{"endpoint"};
    } else {
        if ($main::CMDARGS{"environment"} eq "prod") {
            $url = "https://api.apix.fi/invoices";
        } elsif ($main::CMDARGS{"environment"} eq "test") {
            $url = "https://test-api.apix.fi/invoices";
        }
    }

    my ($code, $response) = ApixPublicApi::SendInvoiceZIP($url, $main::CMDARGS{"software_name"}, $main::CMDARGS{"software_version"}, $main::CMDARGS{"transfer_id"}, $main::CMDARGS{"transfer_key"}, $data);

    if ($code != 200) {
        die("Unknown HTTP error: [${code}] ${response}\n");
    }

    # This is quick-and-dirty solution, real solution would be a use xml parser to get data
    if ($response =~ /<Status>OK<\/Status>/) {
        print "Invoice sent successfully:\n${response}\n";
        exit(0);
    } else {
        print "Invoice was rejected (attach api response for possible questions for Apix):\n${response}\n";
        exit(1);
    }

}

sub check_help {

    my $show_help = 0;

    if (defined($main::CMDARGS{"help"})) {
        $show_help = 1;
    }

    my @required = ("software_name", "software_version", "transfer_id", "transfer_key", "file");
    for my $key (@required) {
        if (!defined($main::CMDARGS{$key})) {
            print STDERR "required argument is missing: $key .\n";
            $show_help = 1;
        }
    }

    if (!defined($main::CMDARGS{"endpoint"}) and !defined($main::CMDARGS{"environment"})) {
        print STDERR "required argument is missing: endpoint OR environment.\n";
        $show_help = 1;
    }

    if (defined($main::CMDARGS{"endpoint"}) and defined($main::CMDARGS{"environment"})) {
        print STDERR "Use only either -endpoint or -environment argument.\n";
        $show_help = 1;
    }

    if (defined($main::CMDARGS{"environment"})) {
        if ($main::CMDARGS{"environment"} !~ /^(test|prod)$/) {
            print STDERR "Allowed arguments for environment is test or prod.\n";
            $show_help = 1;
        }
    }

    if ($main::CMDARGS{"file"} and !-e $main::CMDARGS{"file"}) {
        printf STDERR "File '%s' is missing from path.\n", $main::CMDARGS{"file"};
        $show_help = 1;
    }

    if ($show_help) {
        print STDERR "Usage: $0 -software_name <software_name> -software_version <software_version> -transfer_id <transfer_id> -transfer_key <transfer_key> -file <file> (-endpoint <url> OR -environment <test/prod>\n";
        exit(1);
    }

    return 1;
}

sub getopt_handler {
    my ($opt_name, $opt_value) = @_;
    $main::CMDARGS{$opt_name} = $opt_value;
}

